src/v2/auth.rs

```diff
@@ -1,23 +1,161 @@
+            })?
+            .name("method")
+            .ok_or_else(|| Error::from(format!("method not found in {}", header)))?
+            .as_str();
+
+        let serialized_content = format!(
+            r#"{{ "{}": {{ {} }} }}"#,
+            method,
+            captures
+                .iter()
+                .filter_map(|capture| {
+                    match (
+                        capture.name("key").map(|n| n.as_str().to_string()),
+                        capture.name("value").map(|n| n.as_str().to_string()),
+                    ) {
+                        (Some(key), Some(value)) => Some(format!(r#""{}": "{}""#, key, value)),
+                        _ => None,
+                    }
+                })
+                .collect::<Vec<_>>()
+                .join(", "),
+        );
+
+        // Deserialize the content
+        let mut unsupported_keys = std::collections::HashSet::new();
+        let content: WwwAuthenticateHeaderContent = serde_ignored::deserialize(
+            &mut serde_json::Deserializer::from_str(&serialized_content),
+            |path| {
+                unsupported_keys.insert(path.to_string());
+            },
+        )?;
+
+        if !unsupported_keys.is_empty() {
+            warn!("unsupported keys remaining {:#?}", unsupported_keys);
+        }
+
+        Ok(content)
     }
 }
 
+/// Structured content for the Bearer authentication response header.
+#[derive(Debug, Default, PartialEq, Eq, Deserialize)]
+pub(crate) struct WwwAuthenticateHeaderContentBearer {
+    realm: String,
+    service: Option<String>,
+    scope: Option<String>,
+}
+
+impl WwwAuthenticateHeaderContentBearer {
+    fn auth_ep(&self, scopes: &[&str]) -> String {
+        let service = self
+            .service
+            .as_ref()
+            .map(|sv| format!("?service={}", sv))
+            .unwrap_or_default();
+
+        let scope = scopes.iter().enumerate().fold(
+            if scopes.is_empty() {
+                ""
+            } else if service.is_empty() {
+                "?"
+            } else {
+                "&"
+            }
+            .to_string(),
+            |acc, (i, &s)| {
+                let separator = if i > 1 { "&" } else { "" };
+                acc + separator + "scope=" + s
+            },
+        );
+
+        format!("{}{}{}", self.realm, service, scope)
+    }
+}
+
+/// Structured content for the Basic authentication response header.
+#[derive(Debug, Default, PartialEq, Eq, Deserialize)]
+pub(crate) struct WwwAuthenticateHeaderContentBasic {
+    realm: String,
+}
+
 impl Client {
-    async fn get_token_provider(&self) -> Result<String> {
+    /// Make a request and return the response's www authentication header.
+    async fn get_www_authentication_header(&self) -> Result<HeaderValue> {
         let url = {
             let ep = format!("{}/v2/", self.base_url.clone(),);
             match reqwest::Url::parse(&ep) {
```

Comment:
If you don't mind while re-arranging this function, can this be turned into a `.map_err()?` maybe?